### PR TITLE
fix: remove isolated resolve() for compat with browser distribution

### PR DESCRIPTION
**Pull request by cmorten**
This PR contains:
- [x] bugfix
- [ ] feature
- [x] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers: #3934

### Description

The use of `resolve()` here is incompatible with the implementation of the method for browser usage of Rollup, see <https://github.com/rollup/rollup/blob/master/browser/path.ts#L60>, specifically the code:

```ts
export function resolve(...paths: string[]) {
	let resolvedParts = paths.shift()!.split(/[/\\]/);

        // ... rest of function
}
```

where without at least one argument, this function will throw an `Uncaught TypeError` when trying to call `.split()` on `paths.shift()` which will be `undefined` as `paths` is an empty array.

Given the NodeJS implementation of `resolve()`, which will always fallback to prepending the CWD if unable to construct an absolute path, this PR's refactor is entirely equivalent for NodeJS while also supporting the bespoke implementation for the browser distribution. In essense, for NodeJS `resolve(resolve(), source) === resolve(source)`.

NodeJS docs: https://nodejs.org/api/path.html#path_path_resolve_paths

I'm struggling to find any browser specific tests to cover this use-case, please advise if this is something that is required - would appreciate some guidance where to add!

In a way this side-steps the issue, which is the browser implementation of `resolve()` does not quite mirror the NodeJS one. For that it would need to be able to handle no arguments, but then the expected behaviour would be to return the CWD which doesn't make much sense in the browser context... but perhaps it could at least not throw with something like:

```ts
let resolvedParts = (paths.shift() || "").split(/[/\\]/);
```

regardless, it's implementation detail becomes moot if that use is never invoked.

### CODE REVIEW

**Comment by lukastaegert** [comment]
Thanks for tackling this! I understand that core of the problem is that we do not have tests for functionality specific to the browser bundle. It might be possible to set something up using the ESM version in Node 14, though. I will have a look at this in the next days and possibly push something to your branch if you do not mind.

**Comment by cmorten** [comment]
> Thanks for tackling this! I understand that core of the problem is that we do not have tests for functionality specific to the browser bundle. It might be possible to set something up using the ESM version in Node 14, though. I will have a look at this in the next days and possibly push something to your branch if you do not mind.

Sounds perfect!

**Comment by lukastaegert** [comment]
So I ended up pushing quite a bit more to your branch, including a full test-suite for the browser bundle (though I ended up testing the UMD bundle instead of the ESM bundle).
Only thing I could not make to work satisfactorily is merging coverage between the two test runs. I assume since one is testing the browser bundle that is very different from the commonjs bundle the other is testing, sourcemap positions do not always match and the merged coverage was always much lower than the individual coverage.
But you can check coverage for the browser tests via `npm run test:coverage:browser`.

Also, regular `npm test` is now slower because it is doing the full build, not only the CommonJS build. To get the old performance, run `npm run test:cjs`.

From the content, I kept your change because it looks to me like an overall performance improvement, but ended up entirely replacing the resolveId algorithm for the browser, as everything else in that function also would not apply here as we have no disk access and e.g. cannot "try" different extensions as that again relies on file system operations.

**Review by lukastaegert — APPROVED** [review]

**Comment by cmorten** [comment]
All sounds very sensible - thanks a tonne @lukastaegert for putting down all the effort, introducing a new test suite etc. on my behalf! Quite happy being a passenger in my own PR :yum: I'll be sure to follow precedent and extend the new brower suite in any future contributions (if appropriate).

File: browser/error.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export const throwNoFileSystem = (method: string) => (..._args: any[]): never => {
+	throw Object.assign(
+		new Error(
+			`Cannot access the file system (via "${method}") when using the browser build of Rollup. Make sure you supply a plugin with custom resolveId and load hooks to Rollup.`
+		),
+		{ code: 'NO_FS_IN_BROWSER', url: 'https://rollupjs.org/guide/en/#a-simple-example' }
+	);
+};
```
Comment (cmorten) [inline browser/error.ts:8]: For a minor opportunity for code re-use, we could consider:

```suggestion
import { error } from "../src/utils/error.ts"

export const throwNoFileSystem = (method: string) => (..._args: any[]): never => {
	error({
	  message:
	    `Cannot access the file system (via "${method}") when using the browser build of Rollup. Make sure you supply a plugin with custom resolveId and load hooks to Rollup.`,
		code: 'NO_FS_IN_BROWSER',
		url: 'https://rollupjs.org/guide/en/#a-simple-example'
  });
};
```

REF: https://github.com/rollup/rollup/blob/master/src/utils/error.ts#L13

Though can appreciate that might not be wanting `./broswer/` code to reference outside out it's own directory. That said, I'm fairly certain the browser bundle makes use of this code path via the shared core anyway.

File: browser/path.ts [hunk 2]
```diff
@@ -58,9 +58,13 @@ export function relative(from: string, to: string) {
 }
 
 export function resolve(...paths: string[]) {
-	let resolvedParts = paths.shift()!.split(/[/\\]/);
+	const firstPathSegment = paths.shift();
+	if (!firstPathSegment) {
```
Comment (cmorten) [inline browser/path.ts:62]: 👍

File: browser/error.ts (hunk 1 above)
Comment (lukastaegert) [reply browser/error.ts:8]: Good idea! No, this is totally fine to do.

File: test/browser/samples/renormalizes-external-paths/_config.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+			load(id) {
+				switch (id) {
+					case '/main.js':
+						return `import './nested/dep.js';
+import './ext.js';
+import './nested/nested-ext.js';`;
+					case '/dep.js':
+						return `import './ext.js';
+import './nested/nested-ext.js';`;
+					case '/nested/dep.js':
```
Comment (cmorten) [inline test/browser/samples/renormalizes-external-paths/_config.js:29]: Ci failing with:

```console
1 failing

  1) browser
       renormalizes-external-paths: renormalizes external paths if possible:
     Error: Could not load \nested\dep.js (imported by main.js): Unexpected id \nested\dep.js
      at Object.load (test\browser\samples\renormalizes-external-paths\_config.js:37:13)
      at D:\a\rollup\rollup\dist\rollup.browser.js:11:340192
```

I suspect in normalizing the paths on Windows, `/` -> `\` and these case statements won't work. May need to consider using `join()` or `sep` to ensure cross-OS compatibility (assuming this isn't just a bug and you're expecting the ids to have posix like paths?)

Comment (lukastaegert) [reply test/browser/samples/renormalizes-external-paths/_config.js:29]: I think the bug is using native `path` in the test config. I will fix it!

**Pull request merged**